File: daemon/internal/protocol/card.go
package protocol

// Card is a task on a project's board, as clients see it. A card is known by its opaque ID, and
// by its Key (the project and the number) when a person reads or types it.
type Card struct {
	// ID is the card's opaque id. Routes use it.
	ID string `json:"id"`
	// ProjectID is the short id of the project the card belongs to.
	ProjectID string `json:"projectId"`
	// Number counts the project's cards from 1. It is the "41" in "#41". Two projects each have a
	// card 12, so a list that mixes projects shows the project name too.
	Number int `json:"number"`
	// Key is the project id and the number as text, such as "web-dashboard#12".
	Key string `json:"key"`
	// Title is the card's short name.
	Title string `json:"title"`
	// Body is the longer description the agent starts from. It may be empty.
	Body string `json:"body"`
	// State is where the card is on the board.
	State CardState `json:"state"`
	// Agent is which agent program works on the card.
	Agent AgentKind `json:"agent"`
	// Model is the model the agent uses. Empty means the agent's own default.
	Model string `json:"model"`
	// Thinking is how much the agent thinks. It is null when the card has no thinking setting.
	// The tstype tag makes the generated TypeScript `ThinkingMode | null` and required, because
	// this field is always sent, as null when there is no setting: a pointer without `omitempty`
	// encodes nothing away, so the type must not make it optional.
	Thinking *ThinkingMode `json:"thinking" tstype:"ThinkingMode | null"`
	// PermissionMode is how much the agent may do without asking.
	PermissionMode PermissionMode `json:"permissionMode"`
	// Role is the role's name, such as "Implementer". Phase 5 turns this into a role id; until
	// then it is the text name of a starter role, and it may be empty.
	Role string `json:"role"`
	// Labels are the project's labels on this card, in the order they were added. Never null.
	Labels []Label `json:"labels"`
	// Package is the monorepo package the card works in. Empty when the project is not one.
	Package string `json:"package"`
	// PlannedStart and PlannedEnd are when the card is meant to run (the Timeline draws these for
	// a card that has not started). Null when the person set no plan.
	PlannedStart *Timestamp `json:"plannedStart" tstype:"Timestamp | null"`
	PlannedEnd   *Timestamp `json:"plannedEnd" tstype:"Timestamp | null"`
	// Due is when the card is due. Null when it has no date.
	Due *Timestamp `json:"due" tstype:"Timestamp | null"`
	// ActualStart and ActualEnd are when the card really ran (the Timeline draws these once it
	// has started, decision D1). Null before the card starts, and before it ends.
	ActualStart *Timestamp `json:"actualStart" tstype:"Timestamp | null"`
	ActualEnd   *Timestamp `json:"actualEnd" tstype:"Timestamp | null"`
	// PullRequest is the card's pull request. Null when it has none. Phase 1 has no CI or forge
	// integration, so it is only ever set by hand or by the fixture.
	PullRequest *PullRequest `json:"pullRequest" tstype:"PullRequest | null"`
	// CI is the state of the card's CI run. Null when the card has no CI data (nothing invents
	// one: architecture.md section 11.5 and the design-port deviations list).
	CI *CIState `json:"ci" tstype:"CIState | null"`
	// ContextUsed is how full the agent's context window is, as a percentage from 0 to 100.
	ContextUsed int `json:"contextUsed"`
	// NeedsReason is why the card waits on a person. Null unless State is needs.
	NeedsReason *NeedsReason `json:"needsReason" tstype:"NeedsReason | null"`
	// DoingNow is the agent's one-line "doing now", empty when it is not working.
	DoingNow string `json:"doingNow"`
	// Paused is true while a pause holds the card between turns. Slice G of Phase 2 writes it.
	Paused bool `json:"paused"`
	// Pinned is true while the card is kept from sleeping on its own. Automatic sleep is Phase 5,
	// so today it only records what the person chose.
	Pinned bool `json:"pinned"`
	// Session is the state of the card's session as it was last stored: awake, working, asleep,
	// waking, and the rest. It is null when the card has never had a session. It is the stored
	// state rather than whether a process runs in the daemon right now, so it survives a restart:
	// a card put to sleep reads asleep after the daemon starts again, and one that was awake reads
	// awake until the restore resumes it (docs/architecture.md 5.3). Every change to it is also
	// announced as card.updated on the project topic, beside session.state_changed on the card's
	// own topic. It is always sent, as null when there is no session: a pointer without `omitempty`
	// encodes nothing away, and the `required` flag of the tstype tag makes the generated type
	// `SessionState | null`, where a bare pointer would be generated as an optional field.
	Session *SessionState `json:"session" tstype:"SessionState | null,required"`
	// ViewMode is the view the card's agent runs in: chat, or terminal (docs/architecture.md 4.3).
	// It is stored with the session, so it survives a restart, and it is chat for a card with no
	// session and for one whose session has stopped. It is always sent. Every change to it is
	// announced as card.updated, like the session state it goes with.
	ViewMode CardViewMode `json:"viewMode"`
	// Branch is the Git branch of the card's work. Empty until the card starts.
	Branch string `json:"branch"`
	// CreatedAt is when the card was made.
	CreatedAt Timestamp `json:"createdAt"`
	// UpdatedAt is when the card last changed.
	UpdatedAt Timestamp `json:"updatedAt"`
}

// PullRequest is a card's pull request: the number a person reads and the address it lives at.
type PullRequest struct {
	// Number is the pull request number, such as 287.
	Number int `json:"number"`
	// URL is the address of the pull request. Empty when only the number is known.
	URL string `json:"url"`
}

// NeedsReason is why a card waits on a person: a kind a client can act on, and the sentence the
// app shows.
type NeedsReason struct {
	// Kind is the reason.
	Kind NeedsReasonKind `json:"kind"`
	// Text is the plain sentence shown under the card. It may be empty.
	Text string `json:"text"`
}

// BoardSnapshot is the answer to GET /v1/projects/{id}/board: the columns and every card.
type BoardSnapshot struct {
	// ProjectID is the project the board belongs to.
	ProjectID string `json:"projectId"`
	// Columns are the card states that have a column, in board order. A card in the "merging"
	// state has no column of its own: it is shown in the "ready" column.
	Columns []CardState `json:"columns"`
	// Cards are all the project's cards, in number order.
	Cards []Card `json:"cards"`
	// ServerTime is the daemon's time when the board was made.
	ServerTime Timestamp `json:"serverTime"`
}

// CreateCardRequest is the body of POST /v1/projects/{id}/cards. The card starts in the backlog
// unless StartState names another of the three states a person can add a card in.
type CreateCardRequest struct {
	// Title is the card's name. It cannot be empty.
	Title string `json:"title"`
	// Body is the description. It may be empty.
	Body string `json:"body,omitempty"`
	// Agent is which agent works on the card. Empty means Claude Code.
	Agent AgentKind `json:"agent,omitempty"`
	// Model is the model to use. Empty means the agent's own default.
	Model string `json:"model,omitempty"`
	// Thinking is how much the agent thinks. Empty means no setting.
	Thinking ThinkingMode `json:"thinking,omitempty"`
	// PermissionMode is how much the agent may do without asking. Empty means auto-accept edits.
	PermissionMode PermissionMode `json:"permissionMode,omitempty"`
	// Role is the role's name. Empty means no role.
	Role string `json:"role,omitempty"`
	// Package is the monorepo package. Empty means none.
	Package string `json:"package,omitempty"`
	// StartState is the column the card is added in: backlog (the default when empty), planning,
	// or working. Any other value is refused, because a card reaches the other columns by what
	// happens to it, not by being added there.
	//
	// The card row is written in the backlog and its session is started straight after, so nothing
	// is ever shown as working before its agent exists. If the agent cannot start, the card is
	// removed again and the request fails: adding a card either works or leaves nothing behind.
	StartState CardState `json:"startState,omitempty"`
}

// UpdateCardRequest is the body of PATCH /v1/cards/{id}. A field that is not set is left as it
// is, so a client sends only what the person changed. Labels replaces the whole list when it is
// not null.
type UpdateCardRequest struct {
	// Title renames the card. Null leaves the name.
	Title *string `json:"title,omitempty"`
	// Body replaces the description. Null leaves it.
	Body *string `json:"body,omitempty"`
	// Agent changes the agent program. Null leaves it.
	Agent *AgentKind `json:"agent,omitempty"`
	// Model changes the model. Null leaves it; the empty string clears it.
	Model *string `json:"model,omitempty"`
	// Thinking changes the thinking mode. Null leaves it; the empty string clears it, which is what
	// a model that cannot be told how hard to think leaves on a card.
	Thinking *ThinkingMode `json:"thinking,omitempty" tstype:"ThinkingMode | ''"`
	// PermissionMode changes how much the agent may do. Null leaves it.
	PermissionMode *PermissionMode `json:"permissionMode,omitempty"`
	// Role changes the role name. Null leaves it.
	Role *string `json:"role,omitempty"`
	// Package changes the package. Null leaves it.
	Package *string `json:"package,omitempty"`
	// Labels replaces the card's labels with exactly these ids. Null leaves them.
	Labels *[]string `json:"labels,omitempty"`
	// PlannedStart, PlannedEnd, Due, ActualStart, and ActualEnd set dates. Null leaves the date.
	// A DateChange clears it instead (there is no way to tell "clear" from "absent" with a
	// pointer alone).
	PlannedStart *DateChange `json:"plannedStart,omitempty"`
	PlannedEnd   *DateChange `json:"plannedEnd,omitempty"`
	Due          *DateChange `json:"due,omitempty"`
	ActualStart  *DateChange `json:"actualStart,omitempty"`
	ActualEnd    *DateChange `json:"actualEnd,omitempty"`
	// DoingNow replaces the "doing now" line. Null leaves it.
	DoingNow *string `json:"doingNow,omitempty"`
	// NeedsReason sets or clears why the card waits on a person. Null leaves it.
	NeedsReason *NeedsReason `json:"needsReason,omitempty"`
}

// DateChange sets a card's date, or clears it. It exists because a JSON null and an absent field
// are the same to Go's decoder, so "clear this date" needs its own shape.
type DateChange struct {
	// Clear removes the date. When it is false, At is the new date.
	Clear bool `json:"clear,omitempty"`
	// At is the new date, in UTC with milliseconds, and null when the date is being cleared. It is
	// a pointer because a Timestamp is a struct: `omitempty` cannot leave one out, and a Timestamp
	// that was never set refuses to encode at all.
	At *Timestamp `json:"at,omitempty" tstype:"Timestamp | null"`
}

// MoveCardRequest is the body of POST /v1/cards/{id}/move: the column a person dragged the card
// to. It carries no reason, because the rules of docs/architecture.md 6.1 refuse every manual move
// to Needs you (rule 3): a card waits on a person because an agent is waiting, never because
// somebody dragged it there. The reason a card carries is written by the daemon on its own path,
// through UpdateCardRequest or the session manager.
type MoveCardRequest struct {
	// State is the column to move to. It must be one of the board's columns.
	State CardState `json:"state"`
}

File: apps/web/src/data/mappers/card.ts
import {
  type AgentKind,
  type CardState,
  type CardViewMode,
  type CIState,
  type PermissionMode,
  type SessionState,
  SessionStateAsleep,
  SessionStateAwake,
  SessionStateSleepWarning,
  SessionStateStarting,
  SessionStateWaitingApproval,
  SessionStateWaking,
  SessionStateWorking,
  type ThinkingMode,
  type Card as WireCard,
} from "@marshal/protocol";
import { toMillis } from "./time";

/** A day in milliseconds, for the day numbers the Timeline and the calendar draw. */
const DAY_MS = 86_400_000;

/*
 * A card, from the daemon to the screens and back.
 *
 * The two sides do not speak the same language about four things, and this module is the one place
 * that knows both:
 *
 *   - The daemon names a card by an opaque id for routes and topics, and by its key
 *     (`api#41`) for people. The store is keyed by the key, so the key is the card's id here.
 *   - The daemon sends `claude`, `auto-edits`, and `high`. The screens show "Claude Code",
 *     "Auto-accept edits", and "High".
 *   - The daemon has timestamps; the Timeline and the calendar draw day numbers relative to today.
 *   - The daemon has no dependencies, members, checklists, comments, or cost yet (their own
 *     phases), so a real card carries none, and the parts of a card that own them show their empty
 *     states.
 */

/**
 * How much of the agent's context window a card has used, in percent, which is what the daemon
 * sends. The screens draw a fraction of the window, so a card's `ctx` is that figure over this.
 */
const PERCENT = 100;

/** The agent kinds the daemon sends and the names the screens show. */
const AGENT_NAMES: Record<AgentKind, string> = {
  claude: "Claude Code",
  gemini: "Gemini CLI",
  codex: "Codex",
  builtin: "Built-in agent",
};

/** The opposite direction, for a setting the screens changed. */
const AGENT_KINDS: Record<string, AgentKind> = {
  "Claude Code": "claude",
  "Gemini CLI": "gemini",
  Codex: "codex",
  "Built-in agent": "builtin",
};

/** The permission modes, in the words the screens use. */
const PERM_NAMES: Record<PermissionMode, string> = {
  ask: "Ask",
  "auto-edits": "Auto-accept edits",
  plan: "Plan only",
  "full-auto": "Full auto",
  bypass: "Bypass permissions",
};

const PERM_MODES: Record<string, PermissionMode> = {
  Ask: "ask",
  "Auto-accept edits": "auto-edits",
  "Plan only": "plan",
  "Full auto": "full-auto",
  "Bypass permissions": "bypass",
};

/** The thinking modes, in the words the screens use. */
const THINK_NAMES: Record<ThinkingMode, string> = {
  low: "Low",
  medium: "Medium",
  high: "High",
  "extra-high": "Extra high",
};

const THINK_MODES: Record<string, ThinkingMode> = {
  Low: "low",
  Medium: "medium",
  High: "high",
  "Extra high": "extra-high",
};

/**
 * The day number a moment falls on, counted from the start of the day the daemon is on. The
 * prototype's Timeline and calendar draw day numbers, not dates, and a card's dates come from the
 * daemon as moments.
 */
function dayOf(from: number, at: number | null | undefined): number | null {
  if (!at) return null;
  const start = new Date(from);
  start.setHours(0, 0, 0, 0);
  return Math.round((at - start.getTime()) / DAY_MS);
}

/**
 * What a stored session state means for the card that owns it (section S7c). The card panel, the
 * board, and the Agents view draw one flag for "asleep" and one for "waking", and a card that is
 * being brought back is drawn asleep with a spinner, the way the prototype did, so `asleep` is true
 * for both states and `waking` for the second alone. Every other state, and a card with no session,
 * is neither.
 */
export function sleepFlags(session: SessionState | null | undefined): {
  asleep: boolean;
  waking: boolean;
} {
  return {
    asleep: session === SessionStateAsleep || session === SessionStateWaking,
    waking: session === SessionStateWaking,
  };
}

/**
 * The session states in which a card has a running agent: it is on a turn, it waits for a message,
 * for an approval, or for the idle timer's sleep, or it is being started. Asleep and stopped
 * sessions have no process, a card that never had a session has none either, and a session that is
 * waking has none yet (it becomes awake when the resume has finished), so none of them counts.
 *
 * This is what the daemon's own count of awake cards (`badges.awake`, the live sessions of a
 * project's cards) counts. Home's list of awake cards on the daemon reads stored rows in the
 * awake, working, and waking states instead, and the fixture seeds rows that have no process; those
 * are the daemon's to reconcile, and the screens follow the live sessions here.
 */
const LIVE_SESSIONS: ReadonlySet<SessionState> = new Set<SessionState>([
  SessionStateStarting,
  SessionStateAwake,
  SessionStateWorking,
  SessionStateWaitingApproval,
  SessionStateSleepWarning,
]);

/** True when the state is one in which a card has a running agent. */
export const hasLiveSession = (session: SessionState | null | undefined): boolean =>
  session !== null && session !== undefined && LIVE_SESSIONS.has(session);

/**
 * A card the daemon fills, in the words the screens use for the same things. It is not the mock's
 * `Card`: that type also carries cost, dependencies, members, checklists, comments, the merge
 * percent, and the bypass flag, which the daemon does not speak about yet (their own phases). The
 * sync layer joins the two, which is why this type names only what the daemon owns.
 */
export interface DaemonCard {
  /** The card's key, `<projectId>#<number>`, which is how every screen and map names it. */
  id: string;
  /** The daemon's own id for the card, which its routes take. It is not the key. */
  daemonId: string;
  /** The number within its project, which is what `#41` shows. */
  n: number;
  p: string;
  title: string;
  state: CardState;
  role: string;
  agent: string;
  model: string;
  think: string | null;
  perm: string;
  branch: string | null;
  ci: CIState | null;
  doing: string;
  reason: string;
  pinned: boolean;
  paused: boolean;
  /** The stored state of the card's session, or null when it never had one (section S7c). */
  session: SessionState | null;
  /** The session is asleep or waking: see `sleepFlags`. */
  asleep: boolean;
  /** The session is being resumed. */
  waking: boolean;
  /** Which view the card opens in (section S9): the chat view, or the agent's own terminal. */
  viewMode: CardViewMode;
  pkg: string | null;
  labels: string[];
  s: number | null;
  e: number | null;
  due: number | null;
  ctx: number;
  upd: number;
  pr: number | null;
}

/** A card as the daemon knows it, in the words the screens use for the same things. */
export function toStoredCard(card: WireCard, now: number): DaemonCard {
  return {
    id: card.key,
    daemonId: card.id,
    n: card.number,
    p: card.projectId,
    title: card.title,
    state: card.state,
    role: card.role,
    agent: AGENT_NAMES[card.agent] ?? card.agent,
    model: card.model,
    think: card.thinking ? (THINK_NAMES[card.thinking] ?? null) : null,
    perm: PERM_NAMES[card.permissionMode] ?? card.permissionMode,
    branch: card.branch === "" ? null : card.branch,
    ci: card.ci ?? null,
    doing: card.doingNow,
    reason: card.needsReason?.text ?? "",
    pinned: card.pinned,
    paused: card.paused,
    session: card.session ?? null,
    ...sleepFlags(card.session),
    viewMode: card.viewMode,
    pkg: card.package === "" ? null : card.package,
    labels: card.labels.map((label) => label.name),
    s: dayOf(now, card.plannedStart ? toMillis(card.plannedStart) : null),
    e: dayOf(now, card.plannedEnd ? toMillis(card.plannedEnd) : null),
    due: dayOf(now, card.due ? toMillis(card.due) : null),
    ctx: card.contextUsed / PERCENT,
    upd: toMillis(card.updatedAt),
    pr: card.pullRequest?.number ?? null,
  };
}

/**
 * The fields the daemon fills, so a card that is already in the store is updated in place and only
 * what changed redraws. What names a card (id, p, n) never changes, and the mock-only fields are
 * not here at all, so whatever the app keeps in them is left alone.
 */
export type MirroredCard = Omit<DaemonCard, "id" | "p" | "n">;

/** The daemon's values for the fields it owns on a card the store already has. */
export function mirroredCard(card: WireCard, now: number): MirroredCard {
  const { id: _key, p: _project, n: _number, ...mirrored } = toStoredCard(card, now);
  return mirrored;
}

/** The agent kind a screen's agent name means, or undefined when the catalog does not know it. */
export function agentKindOf(name: string): AgentKind | undefined {
  return AGENT_KINDS[name];
}

/** The permission mode a screen's name means, or undefined when it is not one Marshal knows. */
export function permissionModeOf(name: string): PermissionMode | undefined {
  return PERM_MODES[name];
}

/** The thinking mode a screen's name means, or undefined when it is not one Marshal knows. */
export function thinkingModeOf(name: string): ThinkingMode | undefined {
  return THINK_MODES[name];
}
